### ISSUE TITLE
ValueError: could not broadcast input array from shape (9) into shape (1)

### ISSUE THREAD

**Issue by FabioSeixas** [julienmalard/traDSSAT#18]
**Describe the bug**

Hello again,
Giving sequence on #16, the same code return the following error:

```
Traceback (most recent call last):                                                                                                                      
  File "testando.py", line 9, in <module>                                                                                                               
    x = tradssat.DSSATRun("C:/DSSAT47/Cassava/CRUZ2019.CSX")                                                                                            
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\mgrs\input.py", line 46, in __init__                    
    self.genetics = PeriphGenMgr(crops, cults, cult_levels, model)                                                                                      
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\mgrs\gen_mgr.py", line 15, in __init__                  
    self.files = {                                                                                                                                      
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\mgrs\gen_mgr.py", line 16, in <dictcomp>                
    lvl: GeneticMgr(_get_model(crp, model), cult) for lvl, crp, cult in zip(levels, crops, cultivars)                                                   
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\mgrs\gen_mgr.py", line 51, in __init__                  
    eco_codes = self.cult_file.get_value('ECO#')                                                                                                        
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\file.py", line 104, in get_value                   
    return self._values.get_value(var, sect=sect, subsect=subsect, cond=cond)                                                                           
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\vals.py", line 70, in get_value                    
    return next(s.get_value(var, subsect, cond=cond) for s in sects if var in s)                                                                        
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\vals.py", line 70, in <genexpr>                    
    return next(s.get_value(var, subsect, cond=cond) for s in sects if var in s)                                                                        
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\vals.py", line 241, in get_value                   
    return np.array(val).flatten()                                                                                                                      
ValueError: could not broadcast input array from shape (9) into shape (1)
```

I actualy tried to pass several others experiment files to `DSSATRun` (built-in experiment files) and found other errors:

With `tradssat.DSSATRun("C:/DSSAT47/Maize/EBPL8501.MZX")`:
```
ValueError: could not convert string to float: '720 0'
```

With `tradssat.DSSATRun("C:/DSSAT47/Canola/USON0801.CNX")`

`ValueError: could not convert string to float: 'O 100'`

These last two are built-in experiment files. Anyone with the software have this files in the directory.

**To Reproduce**

1. The same files from #16 for the first description, and default files for the second and third description.

2. Version of traDSSAT: `0.1.5`

3. The necessary code: `tradssat.DSSATRun("C:/DSSAT47/Cassava/CRUZ2019.CSX")`

Thanks

**Comment by julienmalard** [comment]
Hello,
Could you send me the files for the second and third cases? I don't have DSSAT on this particular computer. 
Thank you!

**Comment by julienmalard** [comment]
I just pushed a fix for the first problem described (https://github.com/julienmalard/traDSSAT/commit/3fd29122114de046bdd21ae00edef521855be386). Will wait for the other files to fix the rest and close the issue. In the meantime, you can upgrade with `pip install git+git://github.com/julienmalard/tradssat.git@master`.
Thanks!

**Comment by FabioSeixas** [comment]
Sorry to make you wait. Busy days.


[EBPL8501.txt](https://github.com/julienmalard/traDSSAT/files/4020444/EBPL8501.txt)
[USON0801.txt](https://github.com/julienmalard/traDSSAT/files/4020445/USON0801.txt)

Hope this is enough.

**Comment by julienmalard** [comment]
Hello,
I checked and from what I can see the problem is not in the .EXP file but rather in one of the linked files (likely soil or weather). Can you post the full traceback and also send the linked file mentioned in the traceback (e.g., the .sol. file with the USOT070001 soil if the problem was the soil file, or the USON weather file otherwise)?
Thanks!

**Comment by julienmalard** [comment]
P.S. And no problem for the wait; same here.

**Comment by FabioSeixas** [comment]
The linked files, soil and weather.

[USON0701.txt](https://github.com/julienmalard/traDSSAT/files/4021331/USON0701.txt)
[USON0801.txt](https://github.com/julienmalard/traDSSAT/files/4021332/USON0801.txt)
[US.txt](https://github.com/julienmalard/traDSSAT/files/4021333/US.txt)
[EBCH8401.txt](https://github.com/julienmalard/traDSSAT/files/4021334/EBCH8401.txt)
[EB.txt](https://github.com/julienmalard/traDSSAT/files/4021335/EB.txt)

Thanks!

**Comment by FabioSeixas** [comment]
The traceback. 

Actually, there is some soils files that have comments in its bodies, and don't fit the library standard. That means more tracebacks.

When that file is in the Soil Directory:
[GA.txt](https://github.com/julienmalard/traDSSAT/files/4021340/GA.txt)

```
Traceback (most recent call last):                                                                                                                      
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\file.py", line 199, in _get_var_names              
    name = next(vr for vr in var_names if line.startswith(vr))                                                                                          
StopIteration                                                                                                                                           
                                                                                                                                                        
During handling of the above exception, another exception occurred:                                                                                     
                                                                                                                                                        
Traceback (most recent call last):                                                                                                                      
  File "testando.py", line 8, in <module>                                                                                                               
    tradssat.DSSATRun("C:/DSSAT47/Canola/USON0801.CNX")                                                                                                 
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\mgrs\input.py", line 40, in __init__                    
    self.soil = PeriphSoilMgr(self.exp.get_file_val('ID_SOIL'), field_levels)                                                                           
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\mgrs\soil_mgr.py", line 10, in __init__                 
    self.files = {lvls: SoilMgr(sl) for sl, lvls in zip(soils, levels)}                                                                                 
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\mgrs\soil_mgr.py", line 10, in <dictcomp>               
    self.files = {lvls: SoilMgr(sl) for sl, lvls in zip(soils, levels)}                                                                                 
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\mgrs\soil_mgr.py", line 30, in __init__                 
    file = SoilFile(os.path.join(soils_dir, f))                                                                                                         
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\input.py", line 17, in __init__                    
    super().__init__(file)                                                                                                                              
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\file.py", line 29, in __init__                     
    self._read()                                                                                                                                        
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\file.py", line 44, in _read                        
    self._read_section(section)                                                                                                                         
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\file.py", line 171, in _read_section               
    self._read_subsection(section_name, subblock)                                                                                                       
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\file.py", line 126, in _read_subsection            
    var_names = self._get_var_names(subblock[0])                                                                                                        
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\file.py", line 201, in _get_var_names              
    raise ValueError(                                                                                                                                   
ValueError: No variable matching "J. Boote added 0.1 f" for file GA.SOL.  
```

Then, when that file is in the Soil Directory:
[HC.txt](https://github.com/julienmalard/traDSSAT/files/4021342/HC.txt)

```
raceback (most recent call last):                                                                                                                      
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\file.py", line 199, in _get_var_names              
    name = next(vr for vr in var_names if line.startswith(vr))                                                                                          
StopIteration                                                                                                                                           
                                                                                                                                                        
During handling of the above exception, another exception occurred:                                                                                     
                                                                                                                                                        
Traceback (most recent call last):                                                                                                                      
  File "testando.py", line 8, in <module>                                                                                                               
    tradssat.DSSATRun("C:/DSSAT47/Canola/USON0801.CNX")                                                                                                 
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\mgrs\input.py", line 40, in __init__                    
    self.soil = PeriphSoilMgr(self.exp.get_file_val('ID_SOIL'), field_levels)                                                                           
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\mgrs\soil_mgr.py", line 10, in __init__                 
    self.files = {lvls: SoilMgr(sl) for sl, lvls in zip(soils, levels)}                                                                                 
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\mgrs\soil_mgr.py", line 10, in <dictcomp>               
    self.files = {lvls: SoilMgr(sl) for sl, lvls in zip(soils, levels)}                                                                                 
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\mgrs\soil_mgr.py", line 30, in __init__                 
    file = SoilFile(os.path.join(soils_dir, f))                                                                                                         
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\input.py", line 17, in __init__                    
    super().__init__(file)                                                                                                                              
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\file.py", line 29, in __init__                     
    self._read()                                                                                                                                        
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\file.py", line 44, in _read                        
    self._read_section(section)                                                                                                                         
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\file.py", line 163, in _read_section               
    self._read_subsection(section_name, subblock)                                                                                                       
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\file.py", line 126, in _read_subsection            
    var_names = self._get_var_names(subblock[0])                                                                                                        
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\file.py", line 201, in _get_var_names              
    raise ValueError(                                                                                                                                   
ValueError: No variable matching "SCS Family                                                                                                            
" for file HC.SOL.                            
```


When that file is in the Soil Directory:
[IR.txt](https://github.com/julienmalard/traDSSAT/files/4021346/IR.txt)


```
Traceback (most recent call last):                                                                                                                      
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\file.py", line 199, in _get_var_names              
    name = next(vr for vr in var_names if line.startswith(vr))                                                                                          
StopIteration                                                                                                                                           
                                                                                                                                                        
During handling of the above exception, another exception occurred:                                                                                     
                                                                                                                                                        
Traceback (most recent call last):                                                                                                                      
  File "testando.py", line 8, in <module>                                                                                                               
    tradssat.DSSATRun("C:/DSSAT47/Canola/USON0801.CNX")                                                                                                 
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\mgrs\input.py", line 40, in __init__                    
    self.soil = PeriphSoilMgr(self.exp.get_file_val('ID_SOIL'), field_levels)                                                                           
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\mgrs\soil_mgr.py", line 10, in __init__                 
    self.files = {lvls: SoilMgr(sl) for sl, lvls in zip(soils, levels)}                                                                                 
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\mgrs\soil_mgr.py", line 10, in <dictcomp>               
    self.files = {lvls: SoilMgr(sl) for sl, lvls in zip(soils, levels)}                                                                                 
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\mgrs\soil_mgr.py", line 30, in __init__                 
    file = SoilFile(os.path.join(soils_dir, f))                                                                                                         
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\input.py", line 17, in __init__                    
    super().__init__(file)                                                                                                                              
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\file.py", line 29, in __init__                     
    self._read()                                                                                                                                        
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\file.py", line 44, in _read                        
    self._read_section(section)                                                                                                                         
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\file.py", line 171, in _read_section               
    self._read_subsection(section_name, subblock)                                                                                                       
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\file.py", line 126, in _read_subsection            
    var_names = self._get_var_names(subblock[0])                                                                                                        
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\tmpl\file.py", line 201, in _get_var_names              
    raise ValueError(                                                                                                                                   
ValueError: No variable matching "ALFVG   MVG   NVG WC" for file IR.SOL. 
```

And finally, the error that origin that Issue:
```
Traceback (most recent call last):                                                                                                                      
  File "testando.py", line 8, in <module>                                                                                                               
    tradssat.DSSATRun("C:/DSSAT47/Canola/USON0801.CNX")                                                                                                 
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\mgrs\input.py", line 40, in __init__                    
    self.soil = PeriphSoilMgr(self.exp.get_file_val('ID_SOIL'), field_levels)                                                                           
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\mgrs\soil_mgr.py", line 10, in __init__                 
    self.files = {lvls: SoilMgr(sl) for sl, lvls in zip(soils, levels)}                                                                                 
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Python38\lib\site-packages\tradssat\mgrs\soil_mgr.py", line 10, in <dictcomp>               
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Py levels)}                                                                                     self.soil = PeriphSoilMgr(self.exp.get_file_val('ID_SOIL')thon38\lib\site-packages\tradssat\mgrs\soil_mgr.py", line 30, in __init__                   File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Py                                                                                              self.files = {lvls: SoilMgr(sl) for sl, lvls in zip(soils,thon38\lib\site-packages\tradssat\tmpl\input.py", line 17, in __init__                      File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Py                                                                                              self.files = {lvls: SoilMgr(sl) for sl, lvls in zip(soils,                                                                                          
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Py                                                                                          
    file = SoilFile(os.path.join(soils_dir, f))                                                                                                         
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Py                                                                                          
    super().__init__(file)                                                                                                                              
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Py                                                                                          
    self._read()                                                                                                                                        
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Py                                                                                          
    self._read_section(section)                                                                                                                         
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Py                                                                                          
    section_name, section_lines = self._process_section_header                                                                                          
  File "C:\Users\Fabio Seixas\AppData\Local\Programs\Python\Py                                                                                          
    matr[:] = val                                                                                                                                       
ValueError: could not convert string to float: 'O 100' 
```

**Comment by julienmalard** [comment]
Thanks! Will investigate.

**Comment by julienmalard** [comment]
Sorry for the late follow-up!

Regarding the comments: I think that we can consider that a DSSAT file format error. All comment lines should start with `!` and if they do not, then traDSSAT will certainly have difficulties. I'm not too sure why DSSAT manages to read it, but even if it can, I would say that this probably not something that traDSSAT should be changed to read it.

Regarding the other errors, they are definitively problems in traDSSAT. The one regarding missing `ALFVG` is a missing variable, which I can add easily. Could you let me know what that variable stands for?

Another option would be to rewrite traDSSAT's internals in such a way that it can read all variables present, even if some of them are not explicitly defined...

The last error with `0 100` is a column width error, which will require a bit of work to fix (see discussion on #19).

Thanks!
-Julien

**Comment by julienmalard** [comment]
Closing this for now as I think some other pull requests have fixed these in the meantime. Feel free to reopen if this is not the case!
